### PR TITLE
feature audit

### LINKED ISSUE #38: TODO: migration of OBS Service cargo audit
# Current plans
- [x] Manual source parameters:
  - [x] SrcTar
  - [x] SrcDir
- [ ] Automatic Source Detection by reading `_service` file that contains `cargo_vendor` if manual way is both empty.
  - [x] Read XML and use `cargo_vendor` parameters instead.
- [ ] Check existence of `Cargo.lock`
- [ ] Warn if `cargo_vendor`'s `update` parameter is `false`.

## Miscellaneous Tasks

- [x] Support tags in vendor
- [x] Refactor vendor #40 
- [ ] Use cargo vendor filterer #37

### PR DESCRIPTION
**Pull request by uncomfyhalomacro**
closes #38

### CODE REVIEW

File: cargo/src/bin/cargo_vendor.rs [hunk 1]
```diff
@@ -7,24 +7,6 @@
 // file, You can obtain one at https://mozilla.org/MPL/2.0/.
 // SPDX-License-Identifier: MPL-2.0
 
-#![deny(warnings)]
-#![warn(unused_extern_crates)]
-// Enable some groups of clippy lints.
```
Comment (Firstyear) [inline cargo/src/bin/cargo_vendor.rs:12]: You don't have to remove these, just comment out "deny warnings" and it unblocks you.

**Review by Firstyear — CHANGES_REQUESTED** [review]

File: cargo/src/bin/cargo_vendor.rs (hunk 1 above)
Comment (uncomfyhalomacro) [reply cargo/src/bin/cargo_vendor.rs:12]: ooooh :0

**Review by Firstyear — APPROVED** [review]

**Comment by Firstyear** [comment]
I like the use of the serde xml defs for reading the service file :P

File: cargo/src/audit.rs [hunk 2]
```diff
@@ -73,24 +78,135 @@ pub struct AuditOpts {
...
+}
+
+impl Error for AuditFailed {}
+
 pub trait Audit {
-    fn run_audit(self, opts: &AuditOpts) -> io::Result<()>;
-    fn process_lockfiles(self) -> io::Result<()>;
+    // Run audit sets the workdir before running other stuff
+    fn run_audit(self, opts: &Opts) -> Result<(), AuditFailed>;
+    fn audit_vendored_tar(self, opts: &Opts, workdir: &Path) -> io::Result<()>;
```
Comment (uncomfyhalomacro) [inline cargo/src/audit.rs:119]: You don't need to have opts for this future self, just extract and get the Cargo.lock. You also need to check if there are multiple vendored tarballs. So in this case, use the `glob` crate and use `/**vendor**.tar.**/` and check their lockfiles.

Comment (Firstyear) [reply cargo/src/audit.rs:119]: Ahhh true. That's a good point.

**Review by uncomfyhalomacro — COMMENTED** [review]
Bet

**Pull request merged**